### ISSUE TITLE
Add support for C++

### ISSUE THREAD

**Issue by bartlibert** [kkoomen/vim-doge#2]
I write mostly C++ code, which is documented with [doxygen](http://www.doxygen.nl/).
There are already some doxygen plugins for vim, but none are really that great.
Doxygen also supports some other languages (while not the main focus), so that could be a bonus.

**Comment by kkoomen** [comment]
@bartlibert I had the whole C family on my list but dropped the support for it since quite some people mentioned that they wouldn't add comments using `@param` etc. The max they would do is use a single-line comment to describe briefly what the function does. They wouldn't add anything else.

However, I had some people as well who indeed use Doxygen and prefer to document the code like that. I'll work on it to implement it, since it's up to the user if they want to use it and it's a nice to have for DoGe.

The suggestion I have (in discussion with other C devs who want this support with Doxygen) is to document the function using the following example:

```

/**
 * TODO
 *
 * @param p1 TODO
 * @param p2 TODO
 * @return TODO
 */
```

Would you agree on that? If not, feel free to provide any other suggestions.

**Comment by kkoomen** [comment]
Sorry, accidentally closed the issue. Re-opened it.

**Comment by kkoomen** [comment]
@bartlibert If you can, please provide me with some example test cases that might be "tricky" or quite complex. I know the basic syntax of the C languages, but I am not aware of all the possibilities of how type hints and return types can be defined and how functions can be declared. The general structure I know for a function is `<func-return-type> <func-name>(<func-param-list>) { ... }`. If you can provide more detail then this will definitely help me when writing a good pattern for it.

**Comment by bartlibert** [comment]
I can agree with the format above, it is very similar to what I use now (although I use the "brief" tag as well, but it is not strictly necessary).

As to the "tricky" cases, I'll probably only now they're tricky when I encounter them, but some I can already think of related to functions:

With templates:
`template<class or typename something> <func-return-type> <func-name>(<func-param-list>){ ... }`

New syntax with auto:
`auto <func-name>(<func-param-list>) -> <func-return-type> { ... }`

**Comment by kkoomen** [comment]
I have written a version with 3 quite complex scenarios that I found. These are the generated results:

```
/**
 * TODO
 *
 * @param text TODO
 * @param node TODO
 * @return TODO
 */
void Emitter::append_token(const std::string& text, const AST_Node* node)
{
  //
}

/**
 * TODO
 *
 * @param it_begin TODO
 * @param _end TODO
 * @return TODO
 */
template<typename T>
MyFunc(std::enable_if<is_foreach_iterator<T>, T>::type& it_begin, T& _end) {
  //
}

/**
 * TODO
 *
 * @param a TODO
 * @param b TODO
 * @return TODO
 */
template <class T>
T GetMax (T a, T b) {
  //
}
```

I'll add support  for the `auto` syntax as well.

To confirm: the `template` and `auto` is CPP specific, right? Because I do want to add support for C and Objective-C as well. I assume you have experience in these two languages as well? If so, would you be interested in adding two more issues for C and Objective-C? We can discuss those features in those issues.

**Comment by bartlibert** [comment]
I'm not familiar with Objective-C, so maybe someone else might be better suited to do that.

auto and template are C++-specific as far as I know.

**Comment by kkoomen** [comment]
@bartlibert I just finished the second pattern for PTS functions containing the `auto` syntax. These are the generated results:

```
/**
 * TODO
 *
 * @param it_begin TODO
 * @param _end TODO
 * @return TODO
 */
template<auto n>
auto f(std::enable_if<is_foreach_iterator<T>, T>::type& it_begin, T& _end) -> std::pair<decltype(n), decltype(n)>
{
    return {n, n};
}

/**
 * TODO
 *
 * @param fun TODO
 * @param args TODO
 * @return TODO
 */
template<class F, class... Args>
decltype(auto) PerfectForward(F fun, Args&&... args)
{
  //
}

/**
 * TODO
 *
 * @param i TODO
 * @param args TODO
 * @return TODO
 */
template<class...T> void h(int i = 0, T... args) {
  //
}
```

Is this enough? If not, please provide some more scenario's I should cover. Otherwise if you agree on this I will write tests and merge into develop later on and release a new version later this day.

**Comment by bartlibert** [comment]
I am familiar with C, but only slightly with C# (must have overlooked that one in your list). As for what you have at the moment for C++, that seems fine. Are you ok with new issues when stuff would be missing for C++?
For the "infinite amounts of interfaces", do you mean the "___init" and "__maybe_unused" stuff? Most of the time that are macros that map onto compiler-specific attributes, so they should not be considered for the documentation.

**Comment by kkoomen** [comment]
> As for what you have at the moment for C++, that seems fine. 

Cool, then I'll merge soon.

> Are you ok with new issues when stuff would be missing for C++?

Of course!

> For the "infinite amounts of interfaces", do you mean the "___init" and "__maybe_unused" stuff? Most of the time that are macros that map onto compiler-specific attributes, so they should not be considered for the documentation.

Ah yes, macros. C is not my main language hehe. I'll add C after #2 has been merged.

**Comment by kkoomen** [comment]
@bartlibert I'm seeing different types of files in some repositories with the extension: `ipp` or `hpp`. I'm assuming those filetypes should be supported as well?

**Comment by kkoomen** [comment]
C++ support has been submitted including tests and is merged to master :tada:

**Comment by bartlibert** [comment]
I just realized it does not work for declarations, probably because there the functions end in ";" instead of "{ ... }". Should I create a new issue for this?

**Comment by kkoomen** [comment]
@bartlibert I have re-opened the issue and will work on it immediately. Totally forgot that scenario. I’ll add some tests for it as well!

**Comment by bartlibert** [comment]
I forgot it as well to be honest, so no worries :)

Op vr 28 jun. 2019 om 09:07 schreef Kim Koomen <notifications@github.com>:

**Comment by kkoomen** [comment]
@bartlibert I fixed the scenarios, added tests and merged to master. I even improved the patterns since I came accross this gigantic function (taken from https://github.com/nlohmann/json/blob/develop/include/nlohmann/adl_serializer.hpp#L14-L29)

```cpp
    /*!
    @brief convert a JSON value to any value type
    This function is usually called by the `get()` function of the
    @ref basic_json class (either explicit or via conversion operators).
    @param[in] j        JSON value to read from
    @param[in,out] val  value to write to
    */
    template<typename BasicJsonType, typename ValueType>
    static auto from_json(BasicJsonType&& j, ValueType& val) noexcept(
        noexcept(::nlohmann::from_json(std::forward<BasicJsonType>(j), val)))
    -> decltype(::nlohmann::from_json(std::forward<BasicJsonType>(j), val), void())
    {
        ::nlohmann::from_json(std::forward<BasicJsonType>(j), val);
}
```
so this advanced syntax is also being supported.

Perhaps you can tell me what the `noexcept(noexcept(::nlohmann::from_json(std::forward<BasicJsonType>(j), val)))` part is here. I presume based on the doc comment that the function name is `from_json(BasicJsonType&& j, ValueType& val)` but the "function" that comes after it seems odd to me. Could you explain?

**Comment by kkoomen** [comment]
Merged to master, declarations are supported as well and therefore closed.

**Comment by bartlibert** [comment]
>`noexcept(noexcept(::nlohmann::from_json(std::forward<BasicJsonType>(j), val)))` part is here. I >presume based on the doc comment that the function name is `from_json(BasicJsonType&& j, >ValueType& val)` but the "function" that comes after it seems odd to me. Could you explain?

The name of the function is indeed "from_json". The first noexcept is the noexcept specifier, which indicates that this function does not throw exceptions (https://en.cppreference.com/w/cpp/language/noexcept_spec). The noexcept specifier can take an expression as well, if the condition is only valid in certain cases. The second noexcept is the noexcept operator (https://en.cppreference.com/w/cpp/language/noexcept), which checks if its parameter is noexcept.

So what the code above means is: the `from_json` function won't throw exceptions if the `::nlohmann::from_json` function does not throw exceptions.

**Comment by kkoomen** [comment]
Thank you for your tremendous support! It's been appreciated!
